overlord/snapstate/backend/setup.go

```diff
@@ -325,3 +327,165 @@ func (b Backend) UndoSetupComponent(cpi snap.ContainerPlaceInfo, installRecord *
 func (b Backend) RemoveSnapInhibitLock(instanceName string) error {
 	return runinhibit.RemoveLockFile(instanceName)
 }
+
+// SetupKernelModulesComponents changes kernel-modules configuration by adding
+// compsToInstall. The components currently active are installedComps.
+func (b Backend) SetupKernelModulesComponents(compsToInstall, currentComps []*snap.ComponentSideInfo, ksnapName string, ksnapRev snap.Revision, meter progress.Meter) (err error) {
+	sysd := newSystemd(b.preseed, meter)
+
+	newActiveComps, replaced := mergeCompSideInfos(currentComps, compsToInstall)
+
+	return moveKModsComponentsState(currentComps, newActiveComps,
+		compsToInstall, replaced, ksnapName, ksnapRev, sysd,
+		"after failure to set-up kernel modules components")
+}
+
+// RemoveKernelModulesComponentsSetup changes kernel-modules configuration by
+// removing compsToRemove and making the final state consider only
+// installedComps.
+func (b Backend) RemoveKernelModulesComponentsSetup(compsToRemove, finalComps []*snap.ComponentSideInfo, ksnapName string, ksnapRev snap.Revision, meter progress.Meter) (err error) {
+	sysd := newSystemd(b.preseed, meter)
+
+	currentActiveComps, replaced := mergeCompSideInfos(finalComps, compsToRemove)
+
+	return moveKModsComponentsState(currentActiveComps, finalComps,
+		replaced, compsToRemove, ksnapName, ksnapRev, sysd,
+		"after failure to remove set-up of kernel modules components")
+}
+
+// mergeCompSideInfos returns a merged list from two lists of
+// ComponentSideInfo, and a list with the intersection, using the criteria of
+// the elements having the same ComponentRef. The rest of the data for an
+// element will come from comps2 if ComponentRef if the same in comps1 and
+// comps2.
+func mergeCompSideInfos(comps1, comps2 []*snap.ComponentSideInfo) (merged, instersect []*snap.ComponentSideInfo) {
```

Comment:
"merge" is slightly confusing as the operation is not symmetric, as comps2 wins. We sometimes call this kind of functions "Delta" but maybe is doesn't clearly apply here?